### PR TITLE
Expand size_of docs

### PR DESCRIPTION
**Pull request by Havvy**
This PR does 3 things.

1. Adds a description of what pointer size means to the primitive pages for usize and isize.
2. Says the general size of things is not stable from compiler to compiler.
3. Adds a table of sizes of things that we do guarantee. As this is the first table in the libstd docs, I've included a picture of how that looks.

### CODE REVIEW

**Comment by Havvy** [comment]
The wording in `size_of` really isn't that good. I want help on figuring out how it should be worded, especially since it involves stability guarantees, and I don't know how to properly word that. Perhaps we should say the size is only not guaranteed for `#[repr(Rust)]`?

I don't like the usage of `T` and `TU` in the table. The size of a reference changes depending on whether or not its referring to an unsized object, but I don't know how to put that in the table succinctly. Do we just add `reference size` and point out in the reference module that a reference is pointer size for sized types and twice pointer sized for unsized types?

Are there any other structs that we guarantee the size for? `Box<T>` perhaps? Wrapper structs?

File: src/libcore/mem.rs [hunk 1]
```diff
@@ -180,6 +180,31 @@ pub fn forget<T>(t: T) {
...
+/// i16 | 2
+/// i32 | 4
+/// i64 | 8
+/// usize | pointer size
+/// isize | pointer size
+/// [T, n] | `size_of::<T>() * n`
+/// &T | pointer size
+/// &UT | 2 * pointer size
+/// Option<&T> | pointer size
+/// Option<&U> | 2 * pointer size
```
Comment (QuietMisdreavus) [inline src/libcore/mem.rs]: small nit: `Option<&UT>`

File: src/libcore/mem.rs [hunk 2]
```diff
@@ -180,6 +180,31 @@ pub fn forget<T>(t: T) {
 /// More specifically, this is the offset in bytes between successive
 /// items of the same type, including alignment padding.
```
Comment (scottmcm) [inline src/libcore/mem.rs]: While you're here, consider mentioning that this is "successive elements in an array with that item type", since layout optimization can (in theory at least) change the layout of "successive items of the same type" in a struct.

File: src/libstd/primitive_docs.rs [hunk 3]
```diff
@@ -710,6 +710,10 @@ mod prim_u128 { }
 //
 /// The pointer-sized signed integer type.
 ///
+/// The size of this primitive is how many bytes it takes to reference any
+/// one byte in memory. For example, on a 32 bit target, this is 4 bytes
```
Comment (scottmcm) [inline src/libstd/primitive_docs.rs]: nit: "one byte" here distracted me, though perhaps only in context of reviewing something about sizes.  Perhaps a phrasing like "to reference any location in memory"?

File: src/libcore/mem.rs [hunk 4]
```diff
@@ -180,6 +180,31 @@ pub fn forget<T>(t: T) {
 /// More specifically, this is the offset in bytes between successive
```
Comment (scottmcm) [inline src/libcore/mem.rs]: Perhaps mention that it's always a multiple of `align_of::<T>()`?

File: src/libcore/mem.rs [hunk 5]
```diff
@@ -180,6 +180,31 @@ pub fn forget<T>(t: T) {
...
+/// u64 | 8
+/// i8 | 1
+/// i16 | 2
+/// i32 | 4
+/// i64 | 8
+/// usize | pointer size
+/// isize | pointer size
+/// [T, n] | `size_of::<T>() * n`
+/// &T | pointer size
+/// &UT | 2 * pointer size
```
Comment (scottmcm) [inline src/libcore/mem.rs]: I think the accepted `extern types` RFC means that there will exist types that are not `Sized` but are none-the-less thin pointers.  And there has been discussion of extra-fat pointers for things like `dyn (Debug + Add)`.  So it might be better to not mention this particularly, especially since use cases for relying on it are kinda sketchy.

File: src/libcore/mem.rs [hunk 6]
```diff
@@ -180,6 +180,31 @@ pub fn forget<T>(t: T) {
...
+/// u32 | 4
+/// u64 | 8
+/// i8 | 1
+/// i16 | 2
+/// i32 | 4
+/// i64 | 8
+/// usize | pointer size
+/// isize | pointer size
+/// [T, n] | `size_of::<T>() * n`
+/// &T | pointer size
```
Comment (scottmcm) [inline src/libcore/mem.rs]: Should probably have `*const T` and `*mut T` as well.

File: src/libcore/mem.rs (hunk 5 above)
Comment (Havvy) [reply src/libcore/mem.rs]: So basically the size of `&T` is generally unstable?

File: src/libstd/primitive_docs.rs (hunk 3 above)
Comment (Havvy) [reply src/libstd/primitive_docs.rs]: If I change it to this, I have to also describe what a location in memory is, no?

File: src/libcore/mem.rs (hunk 4 above)
Comment (Havvy) [reply src/libcore/mem.rs]: I don't know what you mean by this.

File: src/libcore/mem.rs (hunk 2 above)
Comment (Havvy) [reply src/libcore/mem.rs]: Done.

File: src/libstd/primitive_docs.rs (hunk 3 above)
Comment (scottmcm) [reply src/libstd/primitive_docs.rs]: Hmm, `*T::offset` talks of "allocated object", so maybe use that instead of location?  Though I don't know if it's defined anywhere (other than LLVM) either...

File: src/libcore/mem.rs [hunk 7]
```diff
@@ -177,15 +177,57 @@ pub fn forget<T>(t: T) {
 
 /// Returns the size of a type in bytes.
 ///
-/// More specifically, this is the offset in bytes between successive
-/// items of the same type, including alignment padding.
+/// More specifically, this is the offset in bytes between successive elements
+/// in an array with that item type including alignment padding. Thus, for any
+/// type `T` and length `n`, [T; n] will have a size of `n * size_of<T>()`.
```
Comment (scottmcm) [inline src/libcore/mem.rs]: `[00:03:45] tidy error: /checkout/src/libcore/mem.rs:182: trailing whitespace`

**Comment by steveklabnik** [comment]
@rust-lang/libs, are you okay with what we are guaranteeing stable here?

File: src/libcore/mem.rs [hunk 8]
```diff
@@ -177,15 +177,57 @@ pub fn forget<T>(t: T) {
...
+/// Pointers to sized types will be the same size as `usize`.
 ///
 /// # Examples
 ///
 /// ```
 /// use std::mem;
 ///
+/// // Some primitives
 /// assert_eq!(4, mem::size_of::<i32>());
+/// assert_eq!(8, mem::size_of::<f64)());
```
Comment (kennytm) [inline src/libcore/mem.rs]: Typo.

```rust
assert_eq!(8, mem::size_of::<f64)());
//                              ^ should be `>`
```

**Comment by sfackler** [comment]
They seem reasonable to me.

File: src/libcore/mem.rs [hunk 9]
```diff
@@ -177,15 +177,57 @@ pub fn forget<T>(t: T) {
...
+/// i64 | 8
+/// f32 | 4
+/// f64 | 8
+/// char | 4
+///
+/// Furthermore, `usize` and `isize` will have the same size.
+///
+/// The types `*T`, `&T`, and `Box<T>` have the same size
+/// and `Option`s of `&T` and `Box<T>` have the same size as the pointer.
+/// Pointers to sized types will be the same size as `usize`.
```
Comment (dtolnay) [inline src/libcore/mem.rs]: Could we say "are" instead of "will be", and "has"/"have" instead of "will have" above? The rest of this documentation is written in present tense so "will" makes it seem like those features are not implemented yet.

File: src/libcore/mem.rs [hunk 10]
```diff
@@ -177,15 +177,57 @@ pub fn forget<T>(t: T) {
...
+/// i32 | 4
+/// i64 | 8
+/// f32 | 4
+/// f64 | 8
+/// char | 4
+///
+/// Furthermore, `usize` and `isize` will have the same size.
+///
+/// The types `*T`, `&T`, and `Box<T>` have the same size
+/// and `Option`s of `&T` and `Box<T>` have the same size as the pointer.
```
Comment (dtolnay) [inline src/libcore/mem.rs]: I found this a little hard to parse. A skeptical reader would read this as:

```
Options    &T
       \  /
        of     Box<T>
          \   /
           and                       the pointer
              \                     /
               have the same size as
```

instead of

```
        &T     Box<T>
          \   /
Options    and
       \  /
        of                       the pointer
          \                     /
           have the same size as
```

because that sounds more plausible to someone skeptical of zero overhead abstractions.

Is there a way to word this that shows the full types `Option<&T>` and `Option<Box<T>>` in the sentence?

Maybe something straightforward like:

> The types *T, &T, Box\<T>, Option\<&T>, and Option\<Box\<T>> all have the same size. If T is [Sized](https://doc.rust-lang.org/std/marker/trait.Sized.html), all of those have the same size as usize.

File: src/libcore/mem.rs [hunk 11]
```diff
@@ -177,15 +177,57 @@ pub fn forget<T>(t: T) {
...
+/// More specifically, this is the offset in bytes between successive elements
+/// in an array with that item type including alignment padding. Thus, for any
+/// type `T` and length `n`, [T; n] will have a size of `n * size_of<T>()`. 
+///
+/// In general, the size of a type is not stable across compilations, but
+/// specific types such as primitives are.
+///
+/// The following table gives the size for primitives.
+///
+/// Type | size_of<Type>()
```
Comment (durka) [inline src/libcore/mem.rs]: `size_of::<Type>()`

File: src/libcore/mem.rs [hunk 12]
```diff
@@ -177,15 +177,57 @@ pub fn forget<T>(t: T) {
...
+/// i16 | 2
+/// i32 | 4
+/// i64 | 8
+/// f32 | 4
+/// f64 | 8
+/// char | 4
+///
+/// Furthermore, `usize` and `isize` will have the same size.
+///
+/// The types `*T`, `&T`, and `Box<T>` have the same size
```
Comment (durka) [inline src/libcore/mem.rs]: This might be the first place we say `*T` instead of `*const T`/`*mut T`? Likewise, everything here about `&T` applies to `&mut T`.

**Review by durka — APPROVED** [review]

File: src/libcore/mem.rs (hunk 7 above)
Comment (durka) [reply src/libcore/mem.rs]: and `size_of::<T>()` here as well

File: src/libcore/mem.rs [hunk 13]
```diff
@@ -177,15 +177,57 @@ pub fn forget<T>(t: T) {
...
+/// assert_eq!(0, mem::size_of::<()>());
+///
+/// // Some arrays
+/// assert_eq!(8, mem::size_of::<[i32; 2]>());
+/// assert_eq!(12, mem::size_of::<[i32, 3]>());
+/// assert_eq!(0, mem::size_of::<[i32, 0]>());
+///
+///
+/// // Pointer size equality
+/// assert_eq!(mem::size_of::<&i32>(), mem::size_of::<*i32>());
```
Comment (durka) [inline src/libcore/mem.rs]: error: expected mut or const in raw pointer type (use `*mut T` or `*const T` as appropriate)

File: src/libcore/mem.rs [hunk 14]
```diff
@@ -177,15 +177,57 @@ pub fn forget<T>(t: T) {
...
+/// assert_eq!(mem::size_of::<&i32>(), mem::size_of::<Box<i32>());
+/// assert_eq!(mem::size_of::<&i32>(), mem::size_of::<Option<&i32>>());
+/// assert_eq!(mem::size_of::<Box<i32>>(), mem::size_of::<Option<Box<32>>>());
```
Comment (durka) [inline src/libcore/mem.rs]: typo `Box<32>`

File: src/libcore/mem.rs (hunk 10 above)
Comment (Havvy) [reply src/libcore/mem.rs]: Yeah, I was sort of floundering with that section. That reads a lot better.

**Comment by dtolnay** [comment]
One more type we guarantee is `extern "C" fn(_) -> _` and `Option<extern "C" fn(_) -> _>` which are guaranteed to be the same size as `_ (*)(_)` in C. I don't know whether we also guarantee the same for `extern "Rust"` or whether the size is guaranteed to be the same as usize.

File: src/libcore/mem.rs [hunk 15]
```diff
@@ -177,15 +177,57 @@ pub fn forget<T>(t: T) {
...
 ///
+/// // Some primitives
 /// assert_eq!(4, mem::size_of::<i32>());
+/// assert_eq!(8, mem::size_of::<f64)());
+/// assert_eq!(0, mem::size_of::<()>());
+///
+/// // Some arrays
+/// assert_eq!(8, mem::size_of::<[i32; 2]>());
+/// assert_eq!(12, mem::size_of::<[i32, 3]>());
+/// assert_eq!(0, mem::size_of::<[i32, 0]>());
```
Comment (matthewjasper) [inline src/libcore/mem.rs]: `;` to separate i32 from the length.

Comment (Havvy) [reply src/libcore/mem.rs]: I thought I was running the doctests; I actually wasn't.

**Comment by Havvy** [comment]
All nits should be fixed. I don't have dtolnay's thing here. I'm wondering if the null pointer optimization for Option-shaped enums is actually stable, and if so, we could just say that, and then point out what the common nullable pointer types are, including `fn` pointers.

I'm avoiding talking about `#[repr(C)]` or any other representation here, since I want to do that in its own PR if we want to talk about representations at all in size_of.

I can also squash all the commits into one if you want.

**Comment by scottmcm** [comment]
I like what you have here; the `Option<&_>` and `Option<Box<_>>` cases are certainly the critical ones.

Details about enum size is probably mixed up in https://github.com/rust-lang/rust/issues/27730 & https://github.com/rust-lang/rfcs/issues/1230, and fine to leave for later.

File: src/libcore/mem.rs [hunk 16]
```diff
@@ -177,15 +177,59 @@ pub fn forget<T>(t: T) {
 
 /// Returns the size of a type in bytes.
 ///
-/// More specifically, this is the offset in bytes between successive
-/// items of the same type, including alignment padding.
+/// More specifically, this is the offset in bytes between successive elements
+/// in an array with that item type including alignment padding. Thus, for any
+/// type `T` and length `n`, [T; n] will have a size of `n * size_of::<T>()`.
```
Comment (dtolnay) [inline src/libcore/mem.rs]: `[T; n]` should be in backticks.

File: src/libcore/mem.rs (hunk 11 above)
Comment (dtolnay) [reply src/libcore/mem.rs]: As durka wrote, this would be clearer as `size_of::<Type>()` instead of `size_of<Type>()`.

File: src/libcore/mem.rs [hunk 17]
```diff
@@ -177,15 +177,59 @@ pub fn forget<T>(t: T) {
...
+/// u64 | 8
+/// i8 | 1
+/// i16 | 2
+/// i32 | 4
+/// i64 | 8
+/// f32 | 4
+/// f64 | 8
+/// char | 4
+///
+/// Furthermore, `usize` and `isize` will have the same size.
```
Comment (dtolnay) [inline src/libcore/mem.rs]: The rest of this documentation is in present tense so "will have" makes it sound like this part is not implemented yet. Plain "have" should be sufficient I think.

File: src/libcore/mem.rs (hunk 8 above)
Comment (dtolnay) [reply src/libcore/mem.rs]: As kennytm wrote, there is a typo after f64.

File: src/libcore/mem.rs [hunk 18]
```diff
@@ -177,15 +177,59 @@ pub fn forget<T>(t: T) {
...
 /// use std::mem;
 ///
+/// // Some primitives
 /// assert_eq!(4, mem::size_of::<i32>());
+/// assert_eq!(8, mem::size_of::<f64)());
+/// assert_eq!(0, mem::size_of::<()>());
+///
+/// // Some arrays
+/// assert_eq!(8, mem::size_of::<[i32; 2]>());
+/// assert_eq!(12, mem::size_of::<[i32, 3]>());
```
Comment (dtolnay) [inline src/libcore/mem.rs]: This should say `[i32; 3]` instead of `[i32, 3]`.

**Review by dtolnay — CHANGES_REQUESTED** [review]

File: src/libcore/mem.rs (hunk 11 above)
Comment (dtolnay) [reply src/libcore/mem.rs]: Also I believe these need to be in backticks or escaped with a backslash. Otherwise markdown will treat it as a \<Type> HTML tag.

```
`Type` | `size_of::<Type>()`
```

```
Type | size_of::\<Type>()
```

**Comment by dtolnay** [comment]
r=me after the last few typos are addressed. I agree we don't need to include `fn` and `repr(C)` in this PR.

**Comment by Havvy** [comment]
Finally figured out how to get rustdoc tests to compile for this. Items in libstd from the libcore facade don't actually get tested from `./x.py test src/libstd` and I have to do `./x.py test src/libcore` instead.

Also rebased everything into a single commit to avoid polluting the commit log with typo fixing and whatnot.

**Review by dtolnay — APPROVED** [review]
